updatesnapyaml - string indices must be integers

I'm not sure what the issue is with swell-foop but I get a [traceback](https://github.com/ubuntu/swell-foop/actions/runs/5073220850/jobs/9112021956) complaining that string indices must be integers:
```
Run ./desktop-snaps/updatesnap/updatesnapyaml.py --github-user $GITHUB_USER --github-token $GITHUB_TOKEN https://github.com/ubuntu/swell-foop
Traceback (most recent call last):
  File "/home/runner/work/swell-foop/swell-foop/./desktop-snaps/updatesnap/updatesnapyaml.py", line 126, in <module>
    main()
  File "/home/runner/work/swell-foop/swell-foop/./desktop-snaps/updatesnap/updatesnapyaml.py", line 95, in main
    parts, tag_error = snap.process_parts()
  File "/home/runner/work/swell-foop/swell-foop/desktop-snaps/updatesnap/SnapModule/snapmodule.py", line 554, in process_parts
    parts.append(self.process_part(part))
  File "/home/runner/work/swell-foop/swell-foop/desktop-snaps/updatesnap/SnapModule/snapmodule.py", line 658, in process_part
    if self._checkopt("allow-neither-tag-nor-branch", version_format):
  File "/home/runner/work/swell-foop/swell-foop/desktop-snaps/updatesnap/SnapModule/snapmodule.py", line 86, in _checkopt
    return (option in dictionary) and dictionary[option]
TypeError: string indices must be integers
Error: Process completed with exit code 1.
```

--- a/updatesnap/SnapModule/snapmodule.py
+++ b/updatesnap/SnapModule/snapmodule.py
@@ -83,6 +83,10 @@ class ProcessVersion:
     def _checkopt(self, option, dictionary):
         """ Returns True if an option is in the dictionary and it's True.
             If it's False or it isn't in the dictionary, it returns False. """
+        if not isinstance(dictionary, dict):
+            raise ValueError(f"The element {dictionary} isn't a dictionary; "
+                             "ensure that your options have the following "
+                             "format: XXXXXXXX: true")
         return (option in dictionary) and dictionary[option]
 
     def _get_version(self, part_name, entry, entry_format, check):
@@ -457,18 +461,33 @@ class Snapcraft(ProcessVersion):
             print(f"Opening file {filename}", file=sys.stderr)
             with open(filename, "r", encoding="utf8") as file_data:
                 data = file_data.read()
-            self._open_yaml_file_with_extensions(data, "updatesnap")
+            if self._open_yaml_file_with_extensions(data, "updatesnap"):
+                self._check_extensions_are_right()
         self._load_secrets(filename)
 
     def load_external_data(self, data, secrets=None):
         """ process SNAPCRAFT.YAML data and SECRETS directly """
 
         self._load_secrets(None)
-        self._open_yaml_file_with_extensions(data, "updatesnap")
+        if self._open_yaml_file_with_extensions(data, "updatesnap"):
+            self._check_extensions_are_right()
         if secrets:
             self._secrets = yaml.safe_load(secrets)
             self._github.set_secrets(self._secrets)
             self._gitlab.set_secrets(self._secrets)
+
+    def _check_extensions_are_right(self):
+        for part_name in self._config["parts"]:
+            part_data = self._config["parts"][part_name]
+            if "version-format" not in part_data:
+                continue
+            if not isinstance(part_data["version-format"], dict):
+                self._print_error(part_name, "The version-format data is not "
+                                  "a dictionary. Ensure that the syntax is right "
+                                  "(specifically, that any option is in the format "
+                                  "'OptionName: true')")
+                raise ValueError(f"The version-format data in {part_name} is not "
+                                 "a dictionary.")
 
     def _open_yaml_file_with_extensions(self, data, ext_name):
         """ This method receives a YAML file content, explores it searching for a comment
@@ -488,6 +507,7 @@ class Snapcraft(ProcessVersion):
 
         newfile = ""
         replace_comments = False
+        has_extensions = False
         for line in data.split("\n"):
             line += '\n'  # restore the newline at the end
             if line[0] != '#':
@@ -496,6 +516,7 @@ class Snapcraft(ProcessVersion):
                 continue
             # the line contains a valid comment
             if line == f'# ext:{ext_name}\n':
+                has_extensions = True
                 replace_comments = True
                 continue
             if line == '# endext\n':
@@ -511,6 +532,7 @@ class Snapcraft(ProcessVersion):
         if data[-1] != '\n' and newfile[-1] == '\n':
             newfile = newfile[:-1]
         self._config = yaml.safe_load(newfile)
+        return has_extensions
 
     def _load_secrets(self, filename):
         secrets_file = os.path.expanduser('~/.config/updatesnap/updatesnap.secrets')
@@ -726,7 +748,6 @@ class Snapcraft(ProcessVersion):
             self._print_message(part, f"  {branch['name']} ({branch['date']})")
 
     def _sort_tags(self, part, current_tag, tags, part_data):
-
         current_date = None
         found_tag = None
         for tag in tags:

--- a/updatesnap/tests/snapcraft_no_true_false.yaml
+++ b/updatesnap/tests/snapcraft_no_true_false.yaml
@@ -1,0 +1,58 @@
+name: swell-foop
+adopt-info: swell-foop
+grade: stable # must be 'stable' to release into candidate/stable channels
+confinement: strict
+base: core22
+
+slots:
+  # for GtkApplication registration
+  swell-foop:
+    interface: dbus
+    bus: session
+    name: org.gnome.SwellFoop
+
+layout:
+ /usr/share/swell-foop:
+   symlink: $SNAP/usr/share/swell-foop
+
+apps:
+  swell-foop:
+    command: bin/snapcraft-preload $SNAP/usr/bin/swell-foop
+    extensions: [gnome]
+    plugs:
+      - opengl
+    desktop: usr/share/applications/org.gnome.SwellFoop.desktop
+    common-id: org.gnome.SwellFoop.desktop
+    environment:
+      GSETTINGS_SCHEMA_DIR: $SNAP/share/glib-2.0/schemas
+      DISABLE_WAYLAND: 'true'
+
+parts:
+  snapcraft-preload:
+# ext:updatesnap
+#   version-format:
+#     allow-neither-tag-nor-branch
+    source: https://github.com/sergiusens/snapcraft-preload.git
+    source-depth: 1
+    plugin: cmake
+    cmake-parameters:
+      - -DCMAKE_INSTALL_PREFIX=/
+    build-packages:
+      - on amd64:
+        - gcc-multilib
+        - g++-multilib
+
+  swell-foop:
+    source: https://gitlab.gnome.org/GNOME/swell-foop.git
+    source-type: git
+    source-tag: '41.1'
+    source-depth: 1
+    plugin: meson
+    meson-parameters:
+      - --prefix=/usr
+    parse-info: [usr/share/metainfo/org.gnome.SwellFoop.appdata.xml]
+    override-build: |
+      sed -i.bak -e 's|Icon=org.gnome.SwellFoop$|Icon=${SNAP}/meta/gui/org.gnome.SwellFoop.png|g' $CRAFT_PART_SRC/data/org.gnome.SwellFoop.desktop.in
+      craftctl default
+      mkdir -p $CRAFT_PART_INSTALL/meta/gui/
+      cp $CRAFT_PART_SRC/data/icons/hicolor/512x512/apps/org.gnome.SwellFoop.png $CRAFT_PART_INSTALL/meta/gui/

--- a/updatesnap/unittests.py
+++ b/updatesnap/unittests.py
@@ -15,6 +15,7 @@ from SnapModule.snapmodule import Gitlab
 
 
 class TestYAMLfiles(unittest.TestCase):
+    # pylint: disable=too-many-public-methods
     """ Unitary tests for snapmodule """
 
     def _load_secrets(self, obj):
@@ -267,7 +268,6 @@ class TestYAMLfiles(unittest.TestCase):
         data = self._base_load_test_file("aligned_comment.yaml")
         yaml_obj = ManageYAML(data)
         element = yaml_obj.get_part_element('gnome-system-monitor', 'source')
-        print(element)
         assert isinstance(element, dict)
         assert 'data' in element
         assert element['data'] == 'source: https://gitlab.gnome.org/GNOME/gnome-system-monitor.git'
@@ -302,6 +302,14 @@ class TestYAMLfiles(unittest.TestCase):
         assert element['level'] == 2
         assert 'separator' in element
         assert element['separator'] == '    '
+
+    def test_no_true_or_false_on_option(self):
+        """ Checks if the file doesn't follow the right format """
+        data = self._base_load_test_file("snapcraft_no_true_false.yaml")
+        snap = Snapcraft(True)
+        with self.assertRaises(ValueError) as context:
+            snap.load_external_data(data)
+        assert context.exception
 
 
 class GitPose:

